update/Pos/Update/Poll/Logic/Apply.hs

```diff
@@ -300,27 +304,32 @@ applyImplicitAgreement (flattenSlotId -> slotId) cd hh = do
 -- confirmed or discarded (approved become confirmed, rejected become
 -- discarded).
 applyDepthCheck
-    :: ApplyMode m
+    :: forall m . ApplyMode m
     => EpochIndex -> HeaderHash -> ChainDifficulty -> m ()
 applyDepthCheck epoch hh (ChainDifficulty cd)
     | cd <= blkSecurityParam = pass
     | otherwise = do
         deepProposals <- getDeepProposals (ChainDifficulty (cd - blkSecurityParam))
+        -- 1. Group psoposals by (application name, difficulty when proposal became decided)
+        -- 2. Sort proposals in each group by tuple
+        --     (decision, whether decision is implicit, positive stake, slot when it has been proposed)
+        -- 3. All proposals in each group except the head we make discarded
+        -- 4. Concat all groups and process all proposals
         let winners =
-                concatMap (toList . resetAllDecisions . NE.sortBy proposalCmp) $
+                concatMap (toList . discardAllExceptHead . NE.sortBy proposalCmp) $
                 NE.groupWith groupCriterion deepProposals
         mapM_ applyDepthCheckDo winners
   where
     upsAppName = svAppName . upSoftwareVersion . upsProposal
-    resetAllDecisions (a:|xs) = a :| map (\x->x {dpsDecision = False}) xs
-    groupCriterion a =
-        ( upsAppName $ dpsUndecided a
-        , dpsDifficulty a)
+    discardAllExceptHead (a:|xs) = a :| map (\x->x {dpsDecision = False}) xs
+    groupCriterion = upsAppName . dpsUndecided
```

Comment:
I suggest that we don't change the criterion unless you can prove that it effectively doesn't change anything (is equivalent to the old one in our case).
Also now it contradicts to the comment you wrote:
> 1. Group psoposals by (application name, difficulty when proposal became decided)

Reply:
as you said in slack: there is one difficulty because otherwise, we processed it on the previous block.

Reply:
I wasn't 100% sure about that, but if you agree (thoughtfully, not just trust me), then ok, let's simplify it (and it looks more correct anyway).

Reply:
I had understood when was writing this code, but wrote in just in case